File: brainfuck/brainfuck.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <stdexcept>
#include <unistd.h>

class BracketNotMatch : public std::exception {
public:
    const char* what() const noexcept override {
        return "The number of matches brackets is incorrect.";
    }
};

class MemoryPosFlow : public std::exception {
public:
    const char* what() const noexcept override {
        return "Memory space has increased too much.";
    }
};

class MemoryPosUnderFlow : public std::exception {
public:
    const char* what() const noexcept override {
        return "Memory location cannot be backward.";
    }
};

class InputisNothing : public std::exception {
public:
    const char* what() const noexcept override {
        return "No inputs.";
    }
};

class BrainFuck
{
private:
    const unsigned int memory_limit = 1 << 20;  // 1MBくらい
    const unsigned int memory_num_limit = 255;
    std::string program;
    std::string inputs;
    int program_pos = 0;
    int cursor_pos = 0;
    std::vector<unsigned char> memory{0, 0, 0};
    std::unordered_map<int, int> bracket_open_map{};
    std::unordered_map<int, int> bracket_close_map{};

    void cursor_pos_increment () {
        if (cursor_pos == memory_limit) {
            throw MemoryPosFlow();
        } else {
            cursor_pos += 1;
            if ((memory.size()-1) < cursor_pos) {
                memory.push_back(0);
            }
        }
    }

    void cursor_pos_decrement () {
        if (cursor_pos == 0) {
            throw MemoryPosUnderFlow();
        } else {
            cursor_pos -= 1;
        }
    }

    void memory_increment () {
        if (memory[cursor_pos] == memory_num_limit) {
            memory[cursor_pos] = 0;
        } else {
            memory[cursor_pos] += 1;
        }
    }

    void memory_decrement () {
        if (memory[cursor_pos] == 0) {
            memory[cursor_pos] = memory_num_limit;
        } else {
            memory[cursor_pos] -= 1;
        }
    }

    void bracket_open () {
        if (memory[cursor_pos] == 0) {
            program_pos = bracket_open_map[program_pos];
        }
    }

    void bracket_close () {
        if (memory[cursor_pos] != 0) {
            program_pos = bracket_close_map[program_pos];
        }
    }

    void memory_out_char () {
        char a = static_cast<char>(memory[cursor_pos]);
        std::cout << a;
    }

    void memory_input_char () {
        // 入力文字がないとき
        if (inputs.empty()) {
            std::getline(std::cin, inputs);
            // 入力文字の中身がなく、EOFの時
            if (inputs.empty() && std::cin.eof()) {
                throw InputisNothing();
            }

        memory[cursor_pos] = static_cast<int>(inputs.front());
        inputs.erase(0);
        }
    }

public:
    explicit BrainFuck(const std::string& n) : program(n) {
        // [と]の数があっているか確認
        int bracket_match = 0;
        std::vector<int> bracket_stash{};

        for (int i=0; i < program.length(); i++) {
            switch (program[i]) {
            case '[':
                bracket_match += 1;
                bracket_stash.push_back(i);
                break;
            case ']':
                bracket_match -= 1;
                if (bracket_match < 0) {
                    throw BracketNotMatch();  // 例えば']['みたいな時
                }
                bracket_open_map.emplace(bracket_stash.back(), i);
                bracket_close_map.emplace(i, bracket_stash.back());
                bracket_stash.pop_back();
                break;
            default:
                break;
            }
        }

        if (bracket_match != 0) {
            throw BracketNotMatch();  // 例えば[][みたいな時
        }
    }

    void execute() {
        while (program_pos < program.length()) {
            switch (program[program_pos]) {
                case '>':
                    cursor_pos_increment();
                    break;
                case '<':
                    cursor_pos_decrement();
                    break;
                case '+':
                    memory_increment();
                    break;
                case '-':
                    memory_decrement();
                    break;
                case '[':
                    bracket_open();
                    break;
                case ']':
                    bracket_close();
                    break;
                case ',':
                    memory_input_char();
                    break;
                case '.':
                    memory_out_char();
                    break;
                default:
                    break;
            }
            program_pos++;
        }
    }
};

int main() {
    std::string input;

    if (isatty(STDIN_FILENO)) {
        // 端末(キーボード)からの実行時
        std::cout << ">>> ";
    }
    std::getline(std::cin, input);

    BrainFuck brain = BrainFuck(input);
    brain.execute();
}

File: brainfuck/brainfuck.py
from sys import stdin
from typing import Final


class BracketNotMatch(Exception):
    def __str__(self):
        return "The number of matches brackets is incorrect."


class MemoryPosFlow(Exception):
    def __str__(self):
        return "Memory space has increased too much."


class MemoryPosUnderFlow(Exception):
    def __str__(self):
        return "Memory location cannot be backward."


class InputisNothing(Exception):
    def __str__(self):
        return "No inputs."


class BrainFuck:
    def __init__(self, program: str):
        # 内部で使う変数の初期化
        self.__MEMORY_LIMIT: Final[int] = (1 << 20)
        self.__MEMORY_NUM_LIMIT: Final[int] = 255
        self.__program: str = program
        self.__inputs: str = ""
        self.__program_pos: int = 0
        self.__cursor_pos: int = 0
        self.__memory: list[int] = [0, 0, 0]
        self.__bracket_open_map = dict()
        self.__bracket_close_map = dict()

        # 一々len関数実行してると実行時間とんでもないことになるので変数で管理する
        self.__memory_length: int = len(self.__memory)

        # ブラケットの確認
        bracket_match: int = 0
        bracket_stash: list[int] = []
        for i in range(len(program)):
            match program[i]:
                case "[":
                    bracket_match += 1
                    bracket_stash.append(i)
                case "]":
                    bracket_match -= 1
                    if bracket_match < 0:
                        raise BracketNotMatch()
                    self.__bracket_open_map[bracket_stash[-1]] = i
                    self.__bracket_close_map[i] = bracket_stash[-1]
                    bracket_stash.pop()

        if bracket_match != 0:
            raise BracketNotMatch()

    def __cursor_pos_increment(self):
        if self.__cursor_pos == self.__MEMORY_LIMIT:
            raise MemoryPosFlow()
        else:
            self.__cursor_pos += 1
            if self.__memory_length-1 < self.__cursor_pos:
                self.__memory.append(0)
                self.__memory_length += 1

    def __cursor_pos_decrement(self):
        if self.__cursor_pos == 0:
            raise MemoryPosUnderFlow()
        else:
            self.__cursor_pos -= 1

    def __memory_increment(self):
        if self.__memory[self.__cursor_pos] == self.__MEMORY_NUM_LIMIT:
            self.__memory[self.__cursor_pos] = 0
        else:
            self.__memory[self.__cursor_pos] += 1

    def __memory_decrement(self):
        if self.__memory[self.__cursor_pos] == 0:
            self.__memory[self.__cursor_pos] = self.__MEMORY_NUM_LIMIT
        else:
            self.__memory[self.__cursor_pos] -= 1

    def __bracket_open(self):
        if self.__memory[self.__cursor_pos] == 0:
            self.__program_pos = self.__bracket_open_map[self.__program_pos]

    def __bracket_close(self):
        if self.__memory[self.__cursor_pos] != 0:
            self.__program_pos = self.__bracket_close_map[self.__program_pos]

    def __memory_out_char(self):
        print(chr(self.__memory[self.__cursor_pos]), end="")

    def __memory_input_char(self):
        if self.__inputs == []:
            try:
                self.__inputs = input()
            except EOFError:
                raise InputisNothing()

        self.__memory[self.__cursor_pos] = ord(self.__inputs[-1])
        self.__inputs[:-1]

    def execute(self):
        program_length = len(self.__program)  # そのままwhile文に入れると毎回len関数呼んで遅くなる
        while self.__program_pos < program_length:
            match self.__program[self.__program_pos]:
                case ">":
                    self.__cursor_pos_increment()
                case "<":
                    self.__cursor_pos_decrement()
                case "+":
                    self.__memory_increment()
                case "-":
                    self.__memory_decrement()
                case "[":
                    self.__bracket_open()
                case "]":
                    self.__bracket_close()
                case ",":
                    self.__memory_input_char()
                case ".":
                    self.__memory_out_char()

            self.__program_pos += 1


if __name__ == "__main__":
    if stdin.isatty():
        print(">>> ", end="")
    inputs = input()
    brainfuck = BrainFuck(inputs)
    brainfuck.execute()
